src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,469 @@
+            }
+        }
+    }
+}
+
+void Publisher::CountUnicastEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    // Count the number of "DNS/SRP Unicast" service entries in
+    // the Network Data.
+
+    const ServiceTlv *serviceTlv = nullptr;
+
+    while ((serviceTlv = Get<Leader>().FindNextMatchingService(
+                serviceTlv, Service::kThreadEnterpriseNumber, &Service::DnsSrpUnicast::kServiceData,
+                sizeof(Service::DnsSrpUnicast::kServiceData))) != nullptr)
+    {
+        TlvIterator      subTlvIterator(*serviceTlv);
+        const ServerTlv *serverSubTlv;
+
+        while (((serverSubTlv = subTlvIterator.Iterate<ServerTlv>())) != nullptr)
+        {
+            if (serviceTlv->GetServiceDataLength() >= sizeof(Service::DnsSrpUnicast::ServiceData))
+            {
+                aNumEntries++;
+
+                // Generally, we prefer entries where the SRP/DNS server
+                // address/port info is included in the service TLV data
+                // over the ones where the info is included in the
+                // server TLV data (i.e., we prefer infra-provided
+                // SRP/DNS entry over a BR local one using MLE-EID). If
+                // our entry itself uses the service TLV data, then we
+                // prefer based on the associated RLOC16.
+
+                if (mType == kTypeUnicast)
+                {
+                    if (IsPreferred(serverSubTlv->GetServer16()))
+                    {
+                        aNumPreferredEntries++;
+                    }
+                }
+                else
+                {
+                    aNumPreferredEntries++;
+                }
+            }
+
+            if (serverSubTlv->GetServerDataLength() >= sizeof(Service::DnsSrpUnicast::ServerData))
+            {
+                aNumEntries++;
+
+                // If our entry also uses the server TLV data (with
+                // ML-EID address), then the we prefer based on the
+                // associated RLOC16.
+
+                if ((mType == kTypeUincastMeshLocalEid) && IsPreferred(serverSubTlv->GetServer16()))
+                {
+                    aNumPreferredEntries++;
+                }
+            }
+        }
+    }
+}
+
+bool Publisher::IsPreferred(uint16_t aRloc16) const
+{
+    // Indicates whether or not an entry from `aRloc16` is preferred
+    // over our entry (based on our RLOC). We prefer an entry from a
+    // router over an entry from an end-device (e.g., a REED). If both
+    // are the same type, then the one with smaller RLOC16 is preferred.
+
+    bool isOtherRouter = Mle::Mle::IsActiveRouter(aRloc16);
+
+    return (Get<Mle::Mle>().IsRouterOrLeader() == isOtherRouter) ? (aRloc16 < Get<Mle::Mle>().GetRloc16())
+                                                                 : isOtherRouter;
+}
+
+void Publisher::UpdateState(uint8_t aNumEntries, uint8_t aNumPreferredEntries, uint8_t aDesiredNumEntries)
```

Comment:
Some rough thoughts for discussion
- `aDesiredNumEntries` (i.e. `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_UNICAST_DNS_SRP_SERVICE_ENTRIES` or `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_ANYCAST_DNS_SRP_SERVICE_ENTRIES`) is configurable, if BRs from different vendors have different configuration, the BR configured with larger value seems always win.  
e.g. if `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_UNICAST_DNS_SRP_SERVICE_ENTRIES` in BR1 is 2, and 3 for BR2 and BR3, BR1 would deregister its service though it might have smaller RLOC16.  Is it expected?
- Should we keep SRP server preferred if it has service records registered already? 
- Should SRP server be notified so that it could remove services registered (and unpublish in advertising proxy) if it re-sign its server role?
- Should SRP client on hold service registration for a while to avoid registering to some server which will be removed?

Reply:
Thanks @librasungirl .

On the publisher configs, we (ideally) want to make sure all BRs behave the same, use the `Publisher` logic with the same set of configs. 

In the last Thread TC call, I brought this up and there was agreement to add this (and the specific configs) as part of Duckhorn BR spec (and may be even cover this as part of cert tests).

----

> Should SRP server be notified so that it could remove services registered (and unpublish in advertising proxy) if it re-sign its server role?

Yes. I think so. I mentioned in one of our internal meeting that as next step we want to add callback to notify next layer and/or SRP server when entry is actually added/removed (to take any action they want). I was hopping/planning to add this as follow up PR but since this PR has been open and under review for rather long time I may end up adding it to the same PR. 😄 

In general, I prefer it if we can all help review PRs more quickly and try to add feature in smaller PRs which build on top of the previous ones. This would be easier to review and also develop and in general I think more agile.

----

> Should we keep SRP server preferred if it has service records registered already?

Good idea. We can add a new API to mark the entry s preferred on device. Again we can add this later in a new PR.

---

> Should SRP client on hold service registration for a while to avoid registering to some server which will be removed?

I would not go for such enhancement unless we have actual data to show there is an issue.

Reply:
> Yes. I think so. I mentioned in one of our internal meeting that as next step we want to add callback to notify next layer and/or SRP server when entry is actually added/removed (to take any action they want).

Why do we need to notify SRP server? I mean are there any actual use case for this feature?

Reply:
> * `aDesiredNumEntries` (i.e. `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_UNICAST_DNS_SRP_SERVICE_ENTRIES` or `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_ANYCAST_DNS_SRP_SERVICE_ENTRIES`) is configurable, if BRs from different vendors have different configuration, the BR configured with larger value seems always win.
>   e.g. if `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_UNICAST_DNS_SRP_SERVICE_ENTRIES` in BR1 is 2, and 3 for BR2 and BR3, BR1 would deregister its service though it might have smaller RLOC16.  Is it expected?

Great point! Seems we should make this feature into the specification?

Reply:
> Great point! Seems we should make this feature into the specification?

@simonlingoogle I did mention that I brought this up ion last Thread Tech committee discussion to be included in the Duckhorn spec (please see https://github.com/openthread/openthread/pull/6587#discussion_r646836099). 


------------

> Why do we need to notify SRP server? I mean are there any actual use case for this feature?

We want server to stop acting as advertising proxy for any previously registered service. Otherwise, the clients that previously registered with it, would notice that its entry is removed fronm Network Data and then try to register with another server which would reject the entries (since old server still proxy advertising the services).

Reply:
Thanks for explanation. @abtink